commercetools-sdk-base/src/main/java/io/sphere/sdk/utils/CompletableFutureUtils.java

```diff
@@ -64,49 +124,178 @@ private CompletableFutureUtils() {
+     * If the recovery function does not require to create a {@link CompletionStage} then use {@link #recover(CompletionStage, Function)}.
+     *
+     * {@include.example io.sphere.sdk.utils.CompletableFutureUtilsTest#recoverWithSuccessInSecond()}
+     *
+     * @param future future which may completes exceptionally
+     * @param f function how the exception should be handled to create a successfully completed future
+     * @param <T> type of the value of the future
+     * @return a future which may recovered from errors
+     */
+    public static <T> CompletableFuture<T> recoverWith(final CompletionStage<T> future, final Function<? super Throwable, CompletionStage<T>> f) {
+        return recoverWith(future, f, ForkJoinPool.commonPool());
+    }
+
+    /**
+     * Creates a {@link CompletionStage} which may recovered from an occurred error by excuting a function in a certain thread pool.
+     * If the recovery does not require to create a {@link CompletionStage} then use {@link #recover(CompletionStage, Function)}.
+     *
+     * Have a look at {@link #recoverWith(CompletionStage, Function)} for an example without the thread pool.
+     *
+     * @param future future which may completes exceptionally
+     * @param f function how the exception should be handled to create a successfully completed future
+     * @param executor thread pool to execute the recover function
+     * @param <T> type of the value of the future
+     * @return a future which may recovered from errors
+     */
+    public static <T> CompletableFuture<T> recoverWith(final CompletionStage<T> future, final Function<? super Throwable, CompletionStage<T>> f, final Executor executor) {
         final CompletableFuture<T> result = new CompletableFuture<>();
         final BiConsumer<T, Throwable> action = (value, error) -> {
             if (value != null) {
                 result.complete(value);
             } else {
                 final CompletionStage<T> alternative = f.apply(error);
-                alternative.whenComplete((alternativeValue, alternativeError) -> {
+                alternative.whenCompleteAsync((alternativeValue, alternativeError) -> {
                     if (alternativeValue != null) {
                         result.complete(alternativeValue);
                     } else {
                         result.completeExceptionally(alternativeError);
                     }
-                });
+                }, executor);
             }
         };
-        future.whenComplete(action);
+        future.whenCompleteAsync(action, executor);
         return result;
     }
 
-    public static <T, X extends Throwable> T orElseThrow(final CompletionStage<T> stage, Supplier<? extends X> exceptionSupplier) throws X, ExecutionException, InterruptedException {
+    /**
+     *
+     * Tries to access the completed value of the future and returns it/or throws the exception inside the future if already available or throws the given exception.
```

Comment:
That's a complex sentence. Let's try:
"Tries to access the completed future if available and returns its value (or exception in case the future completed exceptionally), otherwise throws the given exception." What do you think?